### ISSUE TITLE
Dagger generates new map everytime when using @IntoMap

### ISSUE THREAD

**Issue by anoopnarang** [google/dagger#3905]
I am trying to use @IntoMap feature of dagger.

I have a component with a map
```
@Singleton
@Component(modules = {
        CarModule.class
})
public interface CarComponent {
    Map<Class<?>, Car> carByClass();
}
```

And a module which has lot of providers 
```
@Module
public class CarModule {

    @Provides
    @IntoMap
    @ClassKey(Mercedes.class)
    Car provideMercedes() {
        return new Mercedes();
    }
    
    @Provides
    @IntoMap
    @ClassKey(Audi.class)
    Car provideAudi() {
        return new Audi();
    }
    
    // A lot more car providers here
}

```

To my surprise when dagger generates code for this, it ends up creating a new Immutable map in the method itself.
```
    @Override
    public Map<Class<?>, Car> carByClass() {
      return ImmutableMap.<Class<?>, Car>builderWithExpectedSize(25)
      .put(Mercedes.class, CarModule_ProvideMercedesFactory.provideMercedes(carModule))
      .put(Audi.class, CarModule_ProvideAudiFactory.provideAudi(carModule))
      ...
      ...
     // A lot more map entries here
    }
```

Since I will be using method `carByClass` a lot ( everytime a http request comes in ), it will create this big map again and again, which can give performance problem in future.

Is my understanding correct here ? How to avoid this ?

**Comment by bcorso** [comment]
> Is my understanding correct here ?

Yes, that's correct. We do this because the individual contributions (i.e. the values in the map) may or may not be scoped and if they are not scoped then by definition Dagger needs to create a new one each time the map is injected. If Dagger scoped the map by default then it would indirectly scope all of your contributions as they would all be created only once.

There's currently no way to tell Dagger to scope the map, but if you really want to scope it you can bind it to another binding that is scoped, e.g.

```
@Module
interface CarModule {
    @interface SomeQualifier {} // Create qualifier for multibinding map
  
    @SomeQualifier  // Add qualifier to @IntoMap contribution
    @Provides
    @IntoMap
    @ClassKey(Mercedes.class)
    Car provideMercedes() {
        return new Mercedes();
    }
    
    @SomeQualifier  // Add qualifier to @IntoMap contribution
    @Provides
    @IntoMap
    @ClassKey(Audi.class)
    Car provideAudi() {
        return new Audi();
    }

    ...

    // Bind the qualified multibinding map to an unqualified, scoped version of the map.
    @Binds
    @Singleton // or whichever scope is appropriate for this module
    Map<Class<?>, Provider<Car>> bindMap(@SomeQualifier Map<Class<?>, Provider<Car>> map);
}
```


Also, if all of your `@IntoMap` contributions are in the same module then you shouldn't even be using a multibinding map, just use a normal binding and scope it yourself, e.g.

```
@Module
interface CarModule {
    @Provides
    @Singleton // or whichever scope is appropriate for this module
    Map<Class<?>, Supplier<Car>> provideMercedes() {
        return new ImmutableMap.builder()
            .put(Mercedes.class, () -> new Mercedes())
            .put(Audi.class, () -> new Audi())
            .build();
    }
}
```

**Comment by anoopnarang** [comment]
> If Dagger scoped the map by default then it would indirectly scope all of your contributions as they would all be created only once.

Even if the map is scoped, the mapping is of `Class` -> `Provider<Car>`. 
I am under the impression that `provider.get()` always gives same object if `@singleton` is used else new object everytime. 
What I mean to say here is if map is scoped, that doesn't mean all underlying contributions are scoped as well.

**Comment by anoopnarang** [comment]
Going with first approach, how this @SomeQualifier annotation needs to be defined ?
I am getting `Missing binding` error in component class.

**Comment by anoopnarang** [comment]
Going with the first approach, I am unable to organize the binding and provider.

Because providers need to be in classes and bindings need to be interfaces. 
Will the cross module approach with @SomeQualifier work then ?

**Comment by anoopnarang** [comment]
Ya I deleted that because my original question had
```
    Map<Class<?>, Car> carByClass();
```
instead of 
```
    Map<Class<?>, Provider<Car>> carByClass();
```
which I wasn't aware will work with dagger. I tried later on and found that it works flawlessly, but doesnt solve my original problem.

Just to reiterate the feature request, we should be able to make the map filled up by `@IntoMap` as singleton, as it can be map of provider values as well.

For now, I will try to go with other approaches. Thanks for going through detail on this @bcorso
